dev_swift/05b_early_stopping.ipynb

```diff
@@ -80,65 +118,252 @@
+    "        }\n",
+    "        \n",
+    "        public func plotLRFinder(){\n",
+    "            plot(lrs, losses.map{$0.scalar}, logScale: true, xLabel:\"lr\", yLabel:\"loss\")\n",
+    "        }\n",
+    "        \n",
+    "    }\n",
+    "}\n",
+    "\n",
+    "extension Learner where O.Scalar: PythonConvertible{\n",
+    "    public var recorder: Recorder? {\n",
+    "        for callback in delegates {\n",
+    "            if let recorder = callback as? Recorder { return recorder }\n",
+    "        }\n",
+    "        return nil\n",
+    "    }\n",
+    "}\n",
+    "\n",
+    "/// Hack\n",
+    "extension Learner {\n",
+    "    func makeTrainEvalDelegate() -> TrainEvalDelegate { return TrainEvalDelegate() }\n",
+    "}\n",
+    "\n",
+    "/// A non-generalized learning rate scheduler\n",
+    "extension Learner where O.Scalar == Float {\n",
+    "\n",
+    "    public class ParamScheduler: Delegate {\n",
+    "        public typealias ScheduleFunc = (Float) -> Float\n",
+    "\n",
+    "        // A learning rate schedule from step to float.\n",
+    "        public var scheduler: ScheduleFunc\n",
+    "        \n",
+    "        public init(scheduler: @escaping (Float) -> Float) {\n",
+    "            self.scheduler = scheduler\n",
+    "        }\n",
+    "        \n",
+    "        override public func batchWillStart(learner: Learner) {\n",
+    "            learner.optimizer.learningRate = scheduler(learner.pctEpochs/Float(learner.epochCount))\n",
     "        }\n",
-    "        print(loss)\n",
-    "        optimizer.update(&model[keyPath: variablesKeyPath], along: 𝛁model)\n",
     "    }\n",
     "}"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### LR Finder"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "NB: You may want to also add something that saves the model before running this, and loads it back after running - otherwise you'll lose your weights!"
+   ]
+  },
   {
    "cell_type": "code",
    "execution_count": null,
    "metadata": {},
    "outputs": [],
    "source": [
-    "// Example usage.\n",
-    "var model = Dense<Float>(inputSize: 784, outputSize: 10)\n",
-    "var optimizer = SGD<Dense<Float>, Float>(learningRate: 0.1)\n",
+    "func expSchedule(start: Float, end: Float, pct: Float) -> Float {\n",
+    "    return start * pow(end / start, pct)\n",
+    "}\n",
+    "\n",
+    "func makeAnnealer(start: Float, end: Float, schedule: @escaping (Float, Float, Float) -> Float) -> (Float) -> Float { \n",
+    "    return { pct in return schedule(start, end, pct) }\n",
+    "}\n",
     "\n",
-    "let data = Tensor<Float>(randomNormal: [10, 10, 784])\n",
-    "let labels = Tensor<Float>(randomNormal: [10, 10])\n",
-    "let dataset = Dataset<Example<Float, Float>>(elements: Example<Float, Float>(data: data, labels: labels))\n",
+    "/// LR Finder test\n",
+    "extension Learner where O.Scalar == Float {\n",
```

Comment:
This can be `where O.Scalar: BinaryFloatingPoint`.

Reply:
This one doesn't work FYI, attempting to debug but stuck at the error
```
initializer 'init(_:)' requires that 'Float' conform to 'BinaryInteger'
            learner.optimizer.learningRate = O.Scalar(scheduler(Float(learner.currentIter)/Float(numIter)))
```
If you can iterate on this notebook to make it compile and work with the cells below, please do so.

Reply:
Very odd. I'm using the newest toolchain. Perhaps there is a difference? I noticed that there were some comments from some of the source notebooks that you were waiting on a toolchain update.

Reply:
Ah you're right, updating to the latest toolchain makes this change work.

Reply:
This can be simplified as:
```swift
learner.optimizer.learningRate = scheduler(learner.currentIter / numIter)
```
... when you replace all `Float` with `O.Scalar`.

Reply:
The annealing function doesn't know `O.Scalar` so
```
learner.optimizer.learningRate = scheduler(learner.currentIter / numIter)
```
throws an error. 
```
learner.optimizer.learningRate = O.Scalar(scheduler(learner.currentIter / numIter))
```
doesn't.

Reply:
Ah, I see. My suggestion was based on the assumption that the annealing function was generic over `BinaryFloatingPoint`. I do think just using `Float` is simpler for teaching.

Reply:
Yes, annealing functions as well as metrics can probably be made more generic later on, you're right. Didn't go down there yet.